20240808-custom-limit-email-notifications.md

```diff
@@ -0,0 +1,251 @@
+
+The initial rollout of this feature will be opt-in per project.
+
+### Header
+
+Date: 2024-07-26 
+
+Owner: mobyvb
+
+Accountable:
+- Console Team
+
+Consulted:
+-
+
+Informed:
+-
+
+### Context
+
+Pro users are currently able to set customized project limits for storage and egress. Right now, the only way for a user to check how close they are to their limits is to go to the project dashboard in the satellite UI. A user could also learn about exceeding their limits when uploads or downloads begin failing, but there is no advance notice in this case.
+
+The purpose of "Project Limit Notification Emails" is to inform users automatically when they are getting close to, or exceeding their custom limits.
+
+In addition to notifications about custom limits, we want to notify users when they get close to or exceed their segment limits.
+
+### Goals
+
+1. Give users advance notice that they are getting close to their custom limits (80% used)
+2. Inform users when they have exceeded their custom limits (100% used)
+
+### Approach / Design
+
+**Things we want to avoid**
+* unintentionally sending multiple emails per "threshold event" (e.g. storage usage on project abc has exceeded 80% of the custom limit)
+* failing to send an email for a "threshold event" (e.g. satellite restart at the same time as threshold event, resulting in dropped communications)
+* too long of a delay between occurrence of event and notification (e.g. >30m is probably too long, and 30m may even be too long)
+* significantly impacting metainfo performance
+
+#### Config Updates
+
+A feature flag should be added on the satellite to enable "limit threshold email notifications". It should default to be turned off, until it is turned on in production. This flag enables/disables:
+1. threshold event detection and queueing in the satellite API
+2. threshold event processing and email sending in the satellite core
+
+We need a configuration for the "buffer time" between event occurrence and email sending. This buffer time allows us to deduplicate instances of the same event queued by multiple satellite API pods.
+
+```
+limit-email-notifications.enabled: false
+limit-email-notifications.email-time-buffer: 10m
+```
+
+#### DB Updates
+
+For each project, we need to track whether limit notification emails are enabled for that project. This allows us to implement opting in or out of email notifications on the project settings page of the satellite UI. We could decide to have a single flag to enable/disable all notifications for the project, or add one flag per limit type (storage, bandwidth, segment), providing the user with more granular control.
+
+We also need to track whether emails were sent for each "threshold event":
```

Comment:
@cam-a when we went over this document in design sync, you questioned whether having a flag to indicate whether an email was sent for a particular event was necessary. If I recall correctly, the general reasoning was something like:

> We have some buffer of time that must elapse between when an event is queued and when we send an email (e.g. 10 mins). This allows for deduplication of events across API pods, and if we are deduplicating anyway, we can go ahead and send a single email, forget about it, and due to the buffer time, we should not need to worry about the same threshold event being queued again.

I do think this reasoning is fair in the majority of cases, but there are still some edge cases that make me feel like we should still have event-specific flags. I'm not 100% sure that these are possible, but they seem reasonable to me:
* satellite has multiple API pods, and a disproportionate amount of a project's traffic is being directed to one specific API pod. This pod hits the 80% egress threshold and queues an event. Over 10 minutes elapse, and then the other two pods report that the 80% egress threshold has been exceeded. The user will receive two emails for the same event
* queuing threshold event fails or is skipped for any reason (e.g. db timeout, weird limit state after satellite restart). Having a flag per notification allows us to retry queuing again (although I don't think this case is accounted for in the design doc)

I don't know if these edge cases are likely enough to justify preserving the flags, but I wanted to share these thoughts with you, and check how you feel about keeping vs. removing the notification-specific flags.